src/Field/Field.java

```diff
@@ -2,15 +2,54 @@
 
 public class Field {
 
-
     private final Square[][] field;
 
     public Field(int row, int col) {
         this.field = new Square[row][col];
+        for(int x = 0; x < row; x++) {
+            for(int y = 0; y < col; y++) {
+                field[x][y] = new SafeSquare();
+            }
+        }
     }
 
     public boolean isEmpty() {
-        return true;
+        boolean isEmpty = true;
+        for(int x = 0; x < field.length; x++) {
+            for(int y = 0; y < field[x].length; y++) {
+                if(field[x][y] instanceof MineSquare) {
+                    isEmpty = false;
+                }
+            }
+        }
+        return isEmpty;
+    }
+
+    public int getHeight() {
+        return this.field.length;
+    }
+
+    public int getWidth() {
+        return this.field[0].length;
+    }
+
+    public void placeMineSquare(int row, int col) {
+        field[row][col] = new MineSquare();
+    }
+
+    public Square getSquareAt(int row, int col) {
```

Comment:
This method is ok in this case because your `Square` classes don't have much that you can do it. However in general, this isn't really safe.

The reasoning behind this is similar to the reasoning behind hiding whether you've implemented your field as nested lists or a single list: you should be able to update the internal way you're storing your mines without consumers of the field class knowing about it.

With this method, you're giving the consumer of this class too much control. Here's a kind of contrived example, but hopefully it makes sense as to why this can be dangerous.

```java
// Define a widget class that records how many times it was used
public class MyWidget {
    private int timesUsed;

    public MyWidget() {
        this.timesUsed = 0;
    }

    public int use() {
        this.timesUsed = this.timesUsed + 1;
        return this.timesUsed;
    }
}

// Define a widget holder class that populates an array of widgets and lets you use them and counts the total number of uses
public class WidgetHolder {
    private final MyWidget[] widgets;
    private int totalWidgetUses;

    public WidgetHolder(int numberOfWidgets) {
        totalWidgetUses = 0;
        widgets = new MyWidget[numberOfWidgets];
        for (var i = 0; i < widgets.length; i++) {
            widgets[i] = new MyWidget();
        }
    }

    public int useWidgetAt(int position) {
        totalWidgetUses = totalWidgetUses + 1;
        return this.widgets[position].use();
    }

    // Note: This is the analogue of your slightly dangerous getSquareAt method, this shouldn't be here, we'll see why below
    public MyWidget getWidgetAt(int position) {
        return widgets[position];
    }

    public int getTotalWidgetUses() {
        return totalWidgetUses;
    }
}

// This is our main class where we're going to abuse the fact that we can get access to the underlying widgets
public class Main {

    public static void main(String[] args) {
        var holder = new WidgetHolder(5);

        holder.useWidgetAt(1);
        holder.useWidgetAt(2);
        holder.useWidgetAt(3);

        var totalWidgetUses = holder.getTotalWidgetUses(); // This will have the correct value, 3

        var sneakilyObtainedWidget = holder.getWidgetAt(1);
        sneakilyObtainedWidget.use();

        totalWidgetUses = holder.getTotalWidgetUses(); // This now has the wrong value. It's still 3, because we bypassed the widget holder
    }
}
```

Reply:
I guess the bit I didn't mention is how to get around not having this method in your case, but I think you'll be able to figure that out. Hint: it's similar to what I've done with my `useWidgetAt` method

Reply:
I have updated the method. Is that how it should be?